sklearn/tree/_tree.pyx

```diff
@@ -1872,12 +1872,12 @@ cdef struct BuildPrunedRecord:
     intp_t parent
     bint is_left
 
-cdef _build_pruned_tree(
+cdef void _build_pruned_tree(
     Tree tree,  # OUT
     Tree orig_tree,
     const uint8_t[:] leaves_in_subtree,
     intp_t capacity
-):
+) noexcept:
```

Comment:
So checking https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html, we have:

> If a noexcept function does finish with an exception then it will print a warning message but not allow the exception to propagate further. On the other hand, calling a noexcept function has zero overhead related to managing exceptions, unlike the previous declarations.

But we also have:

> There’s a known performance pitfall when combining nogil and except * @cython.exceptval(check=True). In this case Cython must always briefly re-acquire the GIL after a function call to check if an exception has been raised. This can commonly happen with a function returning nothing (C void). Simple workarounds are to mark the function as noexcept if you’re certain that exceptions cannot be thrown, or to change the return type to int and just let Cython use the return value as an error flag (by default, -1 triggers the exception check).

Which means here we do have the overhead, and it'll print a warning, and keep going and the exceptions raised later in this function are ignored the way things are.